### ISSUE TITLE
Speed issue when importing with SPM

### ISSUE THREAD

**Issue by alessioborraccino** [krzyzanowskim/CryptoSwift#833]
**Describe the bug**
When importing Cryptoswift as a Swift Package, the performance on debug configuration for AESEncrypt/Decrypt is very low.
This is not the case when importing with Carthage. Due to Carthage breaking in Xcode12, it would be good to have the same performance also in SwiftPM if it were possible.

This is probably due to different settings on the project than the default BuildSettings on SwiftPM, but maybe there is way to set the same flags/settings (I wasn't able for now). 

**Reproduce**
Steps to reproduce:
1. Open Package.Swift with Xcode12 
2. Run tests (⌘U)
3. Open CryptoSwift.xcodeProj instead with XCode12
4. Run tests (⌘U)

Tested with CryptoSwift 1.3.2, Xcode12

**Expected**
Performance Times would be similar

**Actual result**
They differ in a very serious amount. 

Package Tests:
```
Test Suite 'AESTestsPerf' started at 2020-10-07 12:05:25.882
Test Case '-[CryptoSwiftTests.AESTestsPerf testAESDecryptPerformance]' started.
/Users/alessio.borraccino/Documents/workspace/CryptoSwift/Tests/CryptoSwiftTests/AESTestsPerf.swift:36: Test Case '-[CryptoSwiftTests.AESTestsPerf testAESDecryptPerformance]' measured [Time, seconds] average: 7.270, relative standard deviation: 1.039%, values: [7.220073, 7.217451, 7.245251, 7.225527, 7.202026, 7.222529, 7.241357, 7.424790, 7.304358, 7.397966], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[CryptoSwiftTests.AESTestsPerf testAESDecryptPerformance]' passed (72.997 seconds).
Test Case '-[CryptoSwiftTests.AESTestsPerf testAESEncryptPerformance]' started.
/Users/alessio.borraccino/Documents/workspace/CryptoSwift/Tests/CryptoSwiftTests/AESTestsPerf.swift:26: Test Case '-[CryptoSwiftTests.AESTestsPerf testAESEncryptPerformance]' measured [Time, seconds] average: 7.905, relative standard deviation: 10.918%, values: [7.092772, 7.097960, 7.133364, 7.189917, 7.801248, 7.990359, 8.930802, 9.650240, 8.736441, 7.431605], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[CryptoSwiftTests.AESTestsPerf testAESEncryptPerformance]' passed (79.314 seconds).
Test Suite 'AESTestsPerf' passed at 2020-10-07 12:07:58.193.
	 Executed 2 tests, with 0 failures (0 unexpected) in 152.311 (152.312) seconds
```

ProjectTests:

```
Test Suite 'AESTestsPerf' started at 2020-10-07 16:12:25.051
Test Case '-[TestsPerformance_iOS.AESTestsPerf testAESDecryptPerformance]' started.
/Users/alessio.borraccino/Documents/workspace/CryptoSwift/Tests/CryptoSwiftTests/AESTestsPerf.swift:36: Test Case '-[TestsPerformance_iOS.AESTestsPerf testAESDecryptPerformance]' measured [Time, seconds] average: 0.082, relative standard deviation: 2.889%, values: [0.088625, 0.080587, 0.081513, 0.081021, 0.080291, 0.082374, 0.080728, 0.080969, 0.080251, 0.081337], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[TestsPerformance_iOS.AESTestsPerf testAESDecryptPerformance]' passed (1.082 seconds).
Test Case '-[TestsPerformance_iOS.AESTestsPerf testAESEncryptPerformance]' started.
/Users/alessio.borraccino/Documents/workspace/CryptoSwift/Tests/CryptoSwiftTests/AESTestsPerf.swift:26: Test Case '-[TestsPerformance_iOS.AESTestsPerf testAESEncryptPerformance]' measured [Time, seconds] average: 0.074, relative standard deviation: 1.670%, values: [0.074480, 0.073003, 0.072712, 0.072978, 0.075217, 0.072736, 0.072409, 0.076419, 0.073530, 0.074271], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[TestsPerformance_iOS.AESTestsPerf testAESEncryptPerformance]' passed (0.995 seconds).
Test Suite 'AESTestsPerf' passed at 2020-10-07 16:12:27.129.
	 Executed 2 tests, with 0 failures (0 unexpected) in 2.077 (2.078) seconds

```

**Comment by krzyzanowskim** [comment]
> the performance on debug configuration for AESEncrypt/Decrypt is very low.

this is a feature of Swift language we can do nothing about. Debug is just many times slower than optimized build.

> This is not the case when importing with Carthage.

Carthage pre-builds binary, and if uses optimized build, that's the result.

> This is probably due to different settings on the project than the default BuildSettings on SwiftPM, but maybe there is way to set the same flags/settings (I wasn't able for now).

Nothing I'm aware of. SwiftPM configuration in Xcode is rather limited when it comes to build customization. Also, SPM comes with only two options "debug" and "release" out of the box. I'll cc @neonichu just in case I'm missing something here.

**Comment by alessioborraccino** [comment]
> Carthage pre-builds binary, and if uses optimized build, that's the result.

That's what i mean, Carthage is using the build settings you set on the Project file, so its optimized and it works as expected. At least that's what i understand. 

> Nothing I'm aware of. SwiftPM configuration in Xcode is rather limited when it comes to build customization. Also, SPM comes with only two options "debug" and "release" out of the box. I'll cc @neonichu just in case I'm missing something here.

It is possible from latest Swift 5.3 to add SwiftSettings to package targets, I tried (unsuccesfully) to add the "-wmo" for debug settings, but I might definitely miss something.

**Comment by krzyzanowskim** [comment]
Turns out I can't distribute `xcframework` with SPM, that would break Linux package as xcframework is for apple platform only. The only thing I can come up now is distribute xcframework separately, and you'd need to copy it over manually, without SPM.

@alessioborraccino you can try this binary (build with Xcode 12.2 on macOS 11 beta 10)
[CryptoSwift.xcframework.zip](https://github.com/krzyzanowskim/CryptoSwift/files/5442053/CryptoSwift.xcframework.zip)

or use [scripts/build-framework.sh](../tree/master/scripts/build-framework.sh) that will build the framework for you.

I'd use your feedback on whether it worked for you.

In the future, I'll attach xcframework to releases on github.

**Comment by alessioborraccino** [comment]
I will ty as soon as i get back to work with it, I have to put it in the backlog for a while. Thanks for your effort, i am looking forward to seeing how the XCFramework works.

**Comment by krzyzanowskim** [comment]
This issue has been addressed by [v1.3.4](https://github.com/krzyzanowskim/CryptoSwift/releases/tag/1.3.4) by distributing prebuild binary `CryptoSwift.xcframework`. This feature requires swift 5.3 (Xcode 12.2).
